dts/arm: stm32h7: quadspi added

SpooniSpoon opened the pull request:
dts/arm: stm32h7: quadspi added

quadspi added to device tree of stm32h7 family.

validated with samples/subsys/fs/littlefs

Signed-off-by: Raphael Löffel <loeffel@rte-ag.ch>

erwango reviewed (commented):
@SpooniSpoon have you been able to validate that this is sufficient to use quadspi on H7 ?

SpooniSpoon commented:
I used samples/subsys/fs/littlefs without DMA support for validation of the dts add-on.

erwango commented:
> I used samples/subsys/fs/littlefs without DMA support for validation of the dts add-on.

Nice! Was it done on an in tree board ?
If yes, would you be able to push related changes (so it can demonstrate how to use it and we could be able to at least compile the driver with H7 configuration in CI) ?

SpooniSpoon commented:
> Nice! Was it done on an in tree board ?
> If yes, would you be able to push related changes (so it can demonstrate how to use it and we could be able to at least compile the driver with H7 configuration in CI) ?

Unfortunately not, I am using a custom board with a W25Q128JVSIM QSPI device.

erwango commented:
> Unfortunately not, I am using a custom board with a W25Q128JVSIM QSPI device.

Ok. From what I can see we have an in-tree board with quad-spi memory (STM32H747I-DISCO),
so we'd need to enable on this board.
Did it work out of the box by just adding this node ? (board configuration put aside)

SpooniSpoon commented:
> Ok. From what I can see we have an in-tree board with quad-spi memory (STM32H747I-DISCO),
> so we'd need to enable on this board.
> Did it work out of the box by just adding this node ? (board configuration put aside)

First of all I had to undefined the DMA node within the QSPI node. After that I still had some compiler errors related to the MDMA. As a quick fix I added `select USE_STM32_HAL_MDMA` to the file `Kconfig.stm32_qspi`. The board configuration I used is

```
&quadspi {
    pinctrl-0 = <&quadspi_clk_pb2
                 &quadspi_bk1_ncs_pb10
                 &quadspi_bk1_io0_pd11
                 &quadspi_bk1_io1_pd12
                 &quadspi_bk1_io2_pe2
                 &quadspi_bk1_io3_pd13>;
//    dmas = <&dma1 5 5 0x0000 0x03>;
//    dma-names = "tx_rx";

    status = "okay";

    w25q128jvsim: qspi-nor-flash@0 {
        compatible = "st,stm32-qspi-nor";
        label = "W25Q128JVSIM";
        reg = <0>;
        qspi-max-frequency = <100000000>;
        /* 128 Megabits = 16 Megabytes */
        size = <0x8000000>;

        status = "okay";

        partitions {
            compatible = "fixed-partitions";
            #address-cells = <1>;
            #size-cells = <1>;

            fs_storage_partition: partition@0 {
                label = "storage";
                reg = <0x00000000 DT_SIZE_M(16)>;
            };
        };
    };
};
```

erwango commented:
> First of all I had to undefined the DMA node within the QSPI node. 

This is not strictly an issue as we can have QPSI working w/o dma.
We'd have QSPI supported on H7 with the limitation that DMA is not yet supported.
But it would need to be clearly specified and documented. Typically generate a dedicated error message 
in case one would try to enable DMA with QSPI on H7.

> After that I still had some compiler errors related to the MDMA. As a quick fix I added select USE_STM32_HAL_MDMA to the file Kconfig.stm32_qspi.

Likely this is because QSPI support requires MDMA enabling in Cube H7 HAL, we can live with that, 
but impacts of having DMA required in HAL and disabled in Zephyr would need a minimum investigation and information to users.


In any case we'd need these both points to be addressed and provided as part of this change so others don't face and report the same issues.

erwango reviewed (changes requested):
Please provide related drivers changes.

SpooniSpoon commented:
> But it would need to be clearly specified and documented.

Where is the right place for documentation in the related yml file of the qspi binding or in the help section of the related Kconfig file?

erwango commented:
> > But it would need to be clearly specified and documented.
> 
> Where is the right place for documentation in the related yml file of the qspi binding or in the help section of the related Kconfig file?

I'd say both Kconfig and yaml.
Also, in the code directly, by generating a dedicated error message in case one would try to enable DMA with QSPI on H7.

FRASTM commented:
commit message `drivers/falsh`  --> `drivers/flash`

erwango commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -321,6 +321,17 @@
 			label = "I2C_4";
 		};
 
+        quadspi: quadspi@52005000 {
+            compatible = "st,stm32-qspi";
+            #address-cells = <1>;
+            #size-cells = <0>;
+            reg = <0x52005000 0x400>;
+            interrupts = <92 0>;
+            clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00004000>;
+            status = "disabled";
+            label = "QUADSPI";
+        };
```
Please fix indentation

erwango reviewed (changes requested):
Please fix commit message for first commit ("quadspi added to device tree of stm32h7 familie")

SpooniSpoon committed:
dts/arm: stm32h7: quadspi added

quadspi added to device tree of stm32h7 family.

validated with samples/subsys/fs/littlefs

Signed-off-by: Raphael Löffel <loeffel@rte-ag.ch>

SpooniSpoon committed:
dts/arm: stm32h7: whitespace issue fixed

whitespace issue fixed to pass compliance checks.

Signed-off-by: Raphael Löffel <loeffel@rte-ag.ch>

erwango commented on drivers/flash/Kconfig.stm32_qspi:
```diff
@@ -18,5 +18,7 @@ config FLASH_STM32_QSPI
 	select FLASH_HAS_PAGE_LAYOUT
 	select DMA if $(DT_STM32_QUADSPI_HAS_DMA)
 	select USE_STM32_HAL_DMA if $(DT_STM32_QUADSPI_HAS_DMA)
+	select USE_STM32_HAL_MDMA if !$(DT_STM32_QUADSPI_HAS_DMA)
```
The proposed logic is not correct.
User always has the possibility of not using DMA wth QSPI, even if supported on the series in use.
With this change, MDMA is systematically enabled in this case.

SpooniSpoon replied:
The MDMA is always required by the STM32 QSPI HAL driver also without DMA support. As a workaround to void compiler errors I enabled the MDMA HAL driver if DMA support is disabled.

erwango replied:
This maybe the case on H7, but not on all series.
You can have a try by doing the folloing change in disco_l475_iot1.dts:
```
 &quadspi {
 	pinctrl-0 = <&quadspi_clk_pe10 &quadspi_ncs_pe11
 		     &quadspi_bk1_io0_pe12 &quadspi_bk1_io1_pe13
 		     &quadspi_bk1_io2_pe14 &quadspi_bk1_io3_pe15>;
-	dmas = <&dma1 5 5 0x0000 0x03>;
-	dma-names = "tx_rx";
+	//dmas = <&dma1 5 5 0x0000 0x03>;
+	//dma-names = "tx_rx";
 
```
Then run:
```
west build -b  disco_l475_iot1 samples/drivers/flash_shell
```

erwango reviewed (changes requested):
One more comment.
Also please squash your 2 commits.

cfriedt commented:
@SpooniSpoon - please rebase and address the last comment

cfriedt commented:
@SpooniSpoon , @erwango - is this one still moving forward?

erwango commented:
@SpooniSpoon Do you plan to continue this PR ?

nvlsianpu reviewed (commented):
similar patch is one of parts of #40859
